File: crates/lib/lib-entity/src/entities/candidate.rs
use chrono::naive::serde::ts_milliseconds_option::serialize as to_milli_tsopt;
use chrono::NaiveDateTime;
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::{company, department, employee, json_schema, position};

#[derive(Copy, Clone, Default, Debug, DeriveEntity)]
pub struct Entity;

impl EntityName for Entity {
    fn table_name(&self) -> &str {
        "candidates"
    }
    fn schema_name(&self) -> Option<&str> {
        None
    }
}

#[derive(Clone, Debug, PartialEq, DeriveModel, DeriveActiveModel, Deserialize, Serialize)]
pub struct Model {
    #[serde(skip)]
    pub id: i32,                          // 主键
    pub company_id: i32,                  // 公司ID
    pub name: String,                     // 候选人姓名
    pub phone: Option<String>,            // 联系电话
    pub email: Option<String>,            // 电子邮箱
    pub position_id: i32,                 // 应聘职位ID
    pub department_id: i32,               // 目标部门ID
    pub interview_date: NaiveDateTime,    // 面试日期
    pub status: Option<String>,           // 状态
    pub interviewer_id: Option<i32>,      // 面试官ID
    pub evaluation: Option<String>,       // 面试评价
    pub remark: Option<String>,           // 备注
    pub extra_value: Option<Value>,       // 额外JSON数据
    pub extra_schema_id: Option<i32>,     // 额外数据schema ID
    #[serde(skip)]
    #[serde(serialize_with = "to_milli_tsopt")]
    pub created_at: NaiveDateTime,        // 创建时间
    #[serde(skip)]
    #[serde(serialize_with = "to_milli_tsopt")]
    pub updated_at: NaiveDateTime,        // 更新时间
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveColumn)]
pub enum Column {
    Id,
    CompanyId,
    Name,
    Phone,
    Email,
    PositionId,
    DepartmentId,
    InterviewDate,
    Status,
    InterviewerId,
    Evaluation,
    Remark,
    ExtraValue,
    ExtraSchemaId,
    CreatedAt,
    UpdatedAt,
}

#[derive(Copy, Clone, Debug, EnumIter, DerivePrimaryKey)]
pub enum PrimaryKey {
    Id,
}

impl PrimaryKeyTrait for PrimaryKey {
    type ValueType = i32;
    fn auto_increment() -> bool {
        true
    }
}

impl ColumnTrait for Column {
    type EntityName = Entity;
    fn def(&self) -> ColumnDef {
        match self {
            Self::Id => ColumnType::Integer.def(),
            Self::CompanyId => ColumnType::Integer.def(),
            Self::Name => ColumnType::String(StringLen::N(64)).def(),
            Self::Phone => ColumnType::String(StringLen::N(20)).def(),
            Self::Email => ColumnType::String(StringLen::N(255)).def().null(),
            Self::PositionId => ColumnType::Integer.def(),
            Self::DepartmentId => ColumnType::Integer.def(),
            Self::InterviewDate => ColumnType::DateTime.def(),
            Self::Status => ColumnType::String(StringLen::N(20)).def(),
            Self::InterviewerId => ColumnType::Integer.def(),
            Self::Evaluation => ColumnType::Text.def().null(),
            Self::Remark => ColumnType::String(StringLen::N(255)).def().null(),
            Self::ExtraValue => ColumnType::Json.def().null(),
            Self::ExtraSchemaId => ColumnType::Integer.def().null(),
            Self::CreatedAt => ColumnType::DateTime
                .def()
                .default(Expr::current_timestamp()),
            Self::UpdatedAt => ColumnType::DateTime
                .def()
                .default(Expr::current_timestamp()),
        }
    }
}

#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {
    Company,
    Position,
    Department,
    Interviewer,
    ExtraSchema,
}

impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        match self {
            Self::Company => Entity::belongs_to(company::Entity)
                .from(Column::CompanyId)
                .to(company::Column::Id)
                .into(),
            Self::Position => Entity::belongs_to(position::Entity)
                .from(Column::PositionId)
                .to(position::Column::Id)
                .into(),
            Self::Department => Entity::belongs_to(department::Entity)
                .from(Column::DepartmentId)
                .to(department::Column::Id)
                .into(),
            Self::Interviewer => Entity::belongs_to(employee::Entity)
                .from(Column::InterviewerId)
                .to(employee::Column::Id)
                .into(),
            Self::ExtraSchema => Entity::belongs_to(json_schema::Entity)
                .from(Column::ExtraSchemaId)
                .to(json_schema::Column::Id)
                .into(),
        }
    }
}

impl ActiveModelBehavior for ActiveModel {} 

File: api/libs/models/candidate.py
from typing import TYPE_CHECKING, Optional
from datetime import datetime
from sqlalchemy import Integer, String, ForeignKey, DateTime, Text, Enum, JSON
from sqlalchemy.orm import Mapped, MappedColumn
from extensions.ext_database import db
from .base import BaseModel
import enum

if TYPE_CHECKING:
    from .json_schema import JsonSchemaInDB
    from .company import CompanyInDB
    from .position import PositionInDB
    from .department import DepartmentInDB
    from .employee import EmployeeInDB


# 候选人状态枚举
class CandidateStatus(str, enum.Enum):
    PENDING = "Pending"  # 待处理
    SCHEDULED = "Scheduled"  # 已安排面试
    INTERVIEWED = "Interviewed"  # 已面试
    ACCEPTED = "Accepted"  # 已通过
    REJECTED = "Rejected"  # 已拒绝
    WITHDRAWN = "Withdrawn"  # 已撤回


class CandidateInDB(BaseModel):
    """候选人信息数据库模型
    存储招聘候选人信息及其与公司、部门、职位的关系

    属性:
        company_id (int): 公司ID，必填
        name (str): 候选人姓名，必填
        phone (str): 联系电话，可选
        email (str): 电子邮箱，可选
        position_id (int): 应聘职位ID，必填
        department_id (int): 目标部门ID，必填
        interview_date (datetime): 面试日期，必填
        status (CandidateStatus): 候选人状态，默认为待处理
        interviewer_id (int): 面试官ID，可选
        evaluation (str): 面试评价，可选
        remark (str): 备注信息，可选
        extra_value (JSON): 附加JSON格式数据，可选
        extra_schema_id (int): 关联的JSON Schema ID，可选

        company (CompanyInDB): 关联的公司对象
        position (PositionInDB): 关联的职位对象
        department (DepartmentInDB): 关联的部门对象
        interviewer (EmployeeInDB): 关联的面试官对象
        extra_schema (JsonSchemaInDB): 关联的JSON Schema对象
    """

    __tablename__ = "candidate"  # 表名

    company_id: Mapped[int] = MappedColumn(
        Integer, ForeignKey("company.id"), nullable=False, comment="公司ID"
    )
    name: Mapped[str] = MappedColumn(String(64), nullable=False, comment="候选人姓名")
    phone: Mapped[Optional[str]] = MappedColumn(
        String(20), nullable=True, comment="联系电话"
    )
    email: Mapped[Optional[str]] = MappedColumn(
        String(255), nullable=True, comment="电子邮箱"
    )
    position_id: Mapped[int] = MappedColumn(
        Integer, ForeignKey("position.id"), nullable=False, comment="应聘职位ID"
    )
    department_id: Mapped[int] = MappedColumn(
        Integer, ForeignKey("department.id"), nullable=False, comment="目标部门ID"
    )
    interview_date: Mapped[datetime] = MappedColumn(
        DateTime, nullable=False, comment="面试日期"
    )
    status: Mapped[CandidateStatus] = MappedColumn(
        Enum(CandidateStatus),
        nullable=False,
        default=CandidateStatus.PENDING,
        comment="候选人状态",
    )
    interviewer_id: Mapped[Optional[int]] = MappedColumn(
        Integer, ForeignKey("employee.id"), nullable=True, comment="面试官ID"
    )
    evaluation: Mapped[Optional[str]] = MappedColumn(
        Text, nullable=True, comment="面试评价"
    )
    remark: Mapped[Optional[str]] = MappedColumn(
        String(255), nullable=True, comment="备注信息"
    )
    extra_value: Mapped[Optional[dict]] = MappedColumn(
        JSON, nullable=True, comment="附加JSON格式数据"
    )
    extra_schema_id: Mapped[Optional[int]] = MappedColumn(
        Integer,
        ForeignKey("json_schemas.id"),
        nullable=True,
        comment="关联的JSON Schema ID",
    )

    # 关系定义
    company: Mapped["CompanyInDB"] = db.relationship(
        "CompanyInDB", foreign_keys=[company_id]
    )  # type: ignore
    position: Mapped["PositionInDB"] = db.relationship(
        "PositionInDB", foreign_keys=[position_id]
    )  # type: ignore
    department: Mapped["DepartmentInDB"] = db.relationship(
        "DepartmentInDB", foreign_keys=[department_id]
    )  # type: ignore
    interviewer: Mapped[Optional["EmployeeInDB"]] = db.relationship(
        "EmployeeInDB", foreign_keys=[interviewer_id]
    )  # type: ignore
    extra_schema: Mapped[Optional["JsonSchemaInDB"]] = db.relationship(
        "JsonSchemaInDB", foreign_keys=[extra_schema_id]
    )  # type: ignore

    def __init__(
        self,
        name: str,
        company_id: int,
        position_id: int,
        department_id: int,
        interview_date: datetime,
        phone: Optional[str] = None,
        email: Optional[str] = None,
        remark: Optional[str] = None,
        evaluation: Optional[str] = None,
        extra_value: Optional[dict] = None,
        interviewer_id: Optional[int] = None,
        extra_schema_id: Optional[int] = None,
        status: CandidateStatus = CandidateStatus.PENDING,
        created_at: Optional[datetime] = None,
        updated_at: Optional[datetime] = None,
    ) -> None:
        self.name = name
        self.company_id = company_id
        self.phone = phone
        self.email = email
        self.position_id = position_id
        self.department_id = department_id
        self.interview_date = interview_date
        self.status = status
        self.interviewer_id = interviewer_id
        self.evaluation = evaluation
        self.remark = remark
        self.extra_value = extra_value
        self.extra_schema_id = extra_schema_id
        self.created_at = created_at or datetime.now()
        self.updated_at = updated_at or datetime.now()
